src/Workspaces/Core/Portable/Shared/Extensions/SafeHandleLease.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions
+{
+    /// <summary>
+    /// Represents a lease of a <see cref="SafeHandle"/>.
+    /// </summary>
+    /// <seealso cref="SafeHandleExtensions.Lease"/>
+    internal readonly struct SafeHandleLease : IDisposable
```

Comment:
maybe a ref struct?  since it should basically always be on the stack ?

Reply:
`ref struct` is a bad pattern for this use case because it prevents use in asynchronous methods. In this case, there are also two cases where it's a field of another handle (essentially child handles which keep the containing handle alive until they are disposed).

Reply:
wfm.